### PR TITLE
Consider timestamp of lastest message as last seen

### PR DESCRIPTION
**Pull request by jgandres**
- Create a Map of userids (key) and last green (value)
- Update last seen with last message ts if it's more recent
- Avoid update last seen with last message ts if we have requested last
seen to server and we don't have received any response

### CODE REVIEW

File: src/presenced.cpp [hunk 1]
```diff
@@ -1272,6 +1307,13 @@ void Client::removePeer(karere::Id peer, bool force)
 
     mCurrentPeers.erase(it);
 
+    // Remove peer from mPeersLastGreen map if exists
+    auto itAux = mPeersLastGreen.find(peer.val);
```
Comment (sergiohs84) [inline src/presenced.cpp]: You can directly do `mPeersLastGreen.erase(peer.val);`, it is equivalent and shorter :)

File: src/presenced.cpp [hunk 2]
```diff
@@ -166,9 +166,44 @@ bool Client::setLastGreenVisible(bool enable)
...
 bool Client::requestLastGreen(Id userid)
 {
+    // Avoid send OP_LASTGREEN if user is ex-contact
+    if (isExContact(userid))
+    {
+        return false;
+    }
+
+    // Reset user last green or insert an entry in the map if not exists
+    std::map<uint64_t, time_t>::iterator it;
```
Comment (sergiohs84) [inline src/presenced.cpp]: All these lines can be simply replaced by `mPeersLastGreen[userid.val] = 0;`. Simplier and shorter :)

File: src/chatd.cpp [hunk 3]
```diff
@@ -1540,6 +1540,8 @@ void Connection::execCommand(const StaticBuffer& buf)
                 }
                 else
                 {
+                    mChatdClient.mKarereClient->updateAndNotifyLastGreen(msg->userid, msg->ts, false);
```
Comment (sergiohs84) [inline src/chatd.cpp]: We don't want to execute this line for every `OLDMSG`, it should be only called `if (opcode == OP_NEWMSG))`

Comment (sergiohs84) [reply src/chatd.cpp]: In any case, probably it's better to call this method once the message is decrypted. Some management messages initially have the `API_USER`/`COMMANDER` as the `userid` and after parsing they change to the user who performed the action (update privileges, change title...) and we would like to update the timestamp in those cases too, if appropriate.

File: src/chatClient.cpp [hunk 4]
```diff
@@ -1145,7 +1145,36 @@ void Client::onPresenceConfigChanged(const presenced::Config& state, bool pendin
 
 void Client::onPresenceLastGreenUpdated(Id userid, uint16_t lastGreen)
 {
-    app.onPresenceLastGreenUpdated(userid, lastGreen);
+    // Calculate lastGreen ts from the system time and format
+    time_t lastGreenTs = time(NULL);
+    lastGreenTs -= (lastGreen * 60);
```
Comment (sergiohs84) [inline src/chatClient.cpp]: I think `time_t lastGreenTs = time(NULL) - (lastGreen * 60);` improves readability

File: src/chatClient.cpp [hunk 5]
```diff
@@ -1145,7 +1145,36 @@ void Client::onPresenceConfigChanged(const presenced::Config& state, bool pendin
...
+    time_t lastMsgTs = 0;
+    SqliteStmt stmt(db, "select MAX(ts) from history where userid = ?");
+    stmt << userid.val;
+    if (stmt.step())
+    {
+        lastMsgTs = stmt.uintCol(0);
+    }
+
+    // Compare lastGreen with last message ts and get the bigger
+    time_t auxLastGreen = (lastGreenTs >= lastMsgTs) ?lastGreenTs :lastMsgTs;
```
Comment (sergiohs84) [inline src/chatClient.cpp]:
```suggestion
    time_t auxLastGreen = (lastGreenTs >= lastMsgTs) ? lastGreenTs : lastMsgTs;
```

**Comment by polmr** [comment]
trigger compilation

**Pull request merged**